duckduckgo/Android
# Allow to UNDO a closed tab

Task/Issue URL: https://app.asana.com/0/1198194956794324/1199173078173336/f
Tech Design URL: 
CC: 

**Description**:
Impement the feature to be able to undo a tab close action in the tab switcher

* DB methods to get tabs filter out "deletable" tabs

* Added specific methods to mark/undo/delete "deletable" tabs in DB, repository and viewmodel classes

* TabSwitcher shows snackbar upon tab close(d) to undo the action

* Register `TabDbSanitizer` app lifecycle observer to clean pending "deletable" tabs upon app launch


**Steps to test this PR**:
Check DB migration
1. update app and relaunch
2. Verify app works normally

Check tab close undo
1. Open several tabs in background
2. Go to tabs screen
3. Close one tab
4. Verify snackbar to undo action pops
5. Let snackbar to auto-dismiss
6. Navitage to main browser screen and back to tabs screen
7. Verify tab is no longer there

Check subsequent tab close undo
1. Open several tabs in background
2. Go to tabs screen
3. Close more than 1 tab before snackbar could auto-dismiss
4. Verify snackbar updates showing the info about the last closed tab
5. Tap "UNDO" CTA in the last snackbar
6. Verify last closed tab reappears
7. Navitage to main browser screen and back to tabs screen
8. Verify tab is no longer there

Check tab (with parent) close undo
1. Open tabs that have a parent tab (long press in link + open in new tab)
2. Go to tabs screen
3. Close parent tab
4. Let snackbar to auto-dismiss
5. Select child tab
6. Navigate back (Android back press)
7. Verify app navigates to empty tab

Check sudden app close
1. Open several tabs in background
2. Go to tabs screen
3. Close one tab and close app before snackbar auto-dismisses (this will leave a "deletable" tab in DB)
4. Open app and navigate to tab switcher
5. Verified closed tab is no longer there

Check selectable tab close
1. Open several tabs in background
2. Go to tabs screen
3. Close the selected tab and let snackbar to auto-dismiss
4. Go back to main browser screen
5. Verify first tab is now showing
6. Repeat steps 1) and 2)
7. Closed the selected tab and immediately go back to browser screen
8. Verify first tab is now showing

Check config changes
1. Open several tabs in background
2. Go to tabs screen
3. Close a tab and immediately rotate the device (portrait -> landscape or vice-versa)
4. Verify the snackbar does not dissapear
5. Undo the close tab action
6. Verify the tab reappears
7. Repeat steps 1) to 3)
8. Let the snackbar auto-dismiss
9. Navigate back to main browser and back into tab switcher
10. Verify the tab is not there

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
